### ISSUE TITLE
[feature-request] emit line and column offsets in json output.

### ISSUE THREAD

**Issue by skywind3000** [languagetool-org/languagetool#991]
[ale (Asynchronous Lint Engine)](https://github.com/w0rp/ale) is a famous vim 8 dedicated plugin which is used widely after vim8 released.

It uses [textlint](https://textlint.github.io/) and [write-good](https://github.com/btford/write-good) for text files linting. and I suppose it is a pity that LanguageTool can't be used in ale.

So, I decided to write ale extension for LanguageTool, the extension must be written in VimScript and will parse the output of LanguageTool.

There are 3 output formats in LanguageTool:

1. readable format (default)
2. json (by --json)
3. xml (by --api)

XML is the best which have all the information I want, but parsing XML is too complex in VimScript, or it requires introducing a lot of dependences. The human readable format missed some key elements such as text range (from which column to which column). 

Parsing json is much easier in VimScript, vim provides json_decode()/json_encode() functions for VimScript. but when I check the actual json output from LanguageTool, I can't find some key like `fromx`, `fromy`, `tox` and `toy` in xml format, the json output just have a `offset` field which is counted by characters.

Can we have the new fields of `fromx`, `fromy`, `tox` and `toy` in json output ? If so, it will be much easier to intergrate languagetool in vimscript.

As we know, parsing json is much easier in many programming language than parsing XML. Including more information in json format will help LanguageTool integrated with other languages.

Do you think it is a reasonable point to include row and line numbers in json output ?

**Comment by danielnaber** [comment]
I chose not to include the information for several reasons:
* it is buggy, i.e. there are off-by-one issues
* I don't see why character based information is not enough - this should include row/line. In the worst case, you could calculate row/line from that value, can't you?

**Comment by skywind3000** [comment]
It is really hard to calculate characters in VimScript:

LanguageTool takes `\r\n` as two characters, but VimScript don't know there is a `\r\n` or a `\n` at the end of line. vim has a `getbufline` api to get lines from a buffer into a list, each line in the list doesn't contain any `\n` or `\r`, I can't figure out the actual line size in bytes.

Could you please just re-consider it ??

**Comment by dpelle** [comment]
If it's buggy, is there any reason we can't fix it in LT?

My LanguageTool vim plugin [1] at...
https://github.com/languagetool-org/languagetool
... uses fromx, fromy, tox, toy values in LT's xml output and
it does not seem to be buggy. I recall that there were bugs
a long time ago but I don't see them anymore.

My LT does not use vim-8 new asynchronous feature, so
the new plugin from Linwei would be welcome.

**Comment by danielnaber** [comment]
I still feel this is more a limitation of vim - an editor that cannot deal with character-based position sounds like a missing feature to me.

**Comment by skywind3000** [comment]
Yes, there is a limitation of vim, could you please help vim users ??

**Comment by danielnaber** [comment]
Can't this be fixed in vim?

**Comment by dpelle** [comment]
I checket the json API at...
https://languagetool.org/http-api/swagger-ui/#!/default/post_check
... and it does not really specify what "offset" is in the response.
If I understand correctly, it's an offset in Unicode character from
the beginning of the document.  The first character would have
offset=0.  Is my interpretation correct?

And even that would still be ambiguous:
* what about composing or combining characters (https://en.wikipedia.org/wiki/Combining_character )
   Are they counted as 1 character or as multiple characters?
* what if the document contains pieces of invalid UTF8? 

> I still feel this is more a limitation of vim - an editor that cannot
> deal with character-based position sounds like a missing feature to me.

Vim has at least the function  byteidx({expr}, {nr}) which
returns the byte index of the {nr}'th character in the string {expr}.
Combining characters are not counted separately (there is also
byteidxcomp(…) to treat combining characters separately.
My vim-LanguageTool plugin uses byteidx(...).

But I agree with @skywind3000 
that having LT provides fromx/tox (etc.) can be more convenient.
The JSON API is meant to be used in various softwares
(vim, emacs, etc.) so it should be convenient to use. 
If only XML has fromx/tox then it's not an incentive
either to use the newer JSON format of LT.  Users will likely
stick with the old XML format.

**Comment by danielnaber** [comment]
You have to click on the "Model" link to get a more detailed description. I'm not sure about combining characters. I'd assume that invalid UTF-8 causes a Java exception, but I haven't tested it.

**Comment by vigoux** [comment]
I ran accross the same issue with `offset` not being really usable in VimScript, and the following solution seems to work for me, it actually places the cursor at the start of the error : 
```vim
execute 'go ' . (byteidx(system("cat " . expand('%')), offset) + 1)
```

From there you can get the starting line and column
Gathering the end point is similar, but `offset` becomes `offset + length - 1`

**Comment by jcs090218** [comment]
Just want to say, same here in Emacs.

**Comment by Britaliope** [comment]
Just to keep you updated, this issue is still blocking the vim integration of LanguageTool to work with LanguageTool 6.0 and higher. I don't know the status of this for emacs.
